flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferBuilder;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static java.lang.Math.addExact;
+import static java.lang.Math.min;
+
+interface ChannelStateSerializer {
+
+	void writeHeader(DataOutputStream dataStream) throws IOException;
+
+	void writeData(DataOutputStream stream, Buffer... flinkBuffers) throws IOException;
+}
+
+interface ChannelStateDeserializer {
+
+	void readHeader(InputStream stream) throws IOException;
+
+	/**
+	 * Reads the length of state.
+	 *
+	 * @return size of state in bytes and offset of the current position resulted from reading.
+	 */
+	Tuple2<Integer, Integer> readLength(InputStream stream) throws IOException;
+
+	int readData(InputStream stream, ChannelStateByteBuffer buffer, int bytes) throws IOException;
+}
+
+/**
+ * Wrapper around various buffers to receive channel state data.
+ */
+@Internal
+@NotThreadSafe
+interface ChannelStateByteBuffer {
+
+	boolean isWritable();
+
+	int writableBytes();
+
+	int writeBytes(InputStream input, int bytesToRead) throws IOException;
+
+	static ChannelStateByteBuffer wrap(Buffer buffer) {
+		return new ChannelStateByteBuffer() {
+
+			private ByteBuf byteBuf = buffer.asByteBuf();
+
+			@Override
+			public boolean isWritable() {
+				return byteBuf.isWritable();
+			}
+
+			@Override
+			public int writableBytes() {
+				return byteBuf.writableBytes();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				return byteBuf.writeBytes(input, Math.min(bytesToRead, byteBuf.writableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(BufferBuilder bufferBuilder) {
+		final byte[] buf = new byte[1024];
```

Comment:
TBH I am concerning of creating the temporary byte array for every buffer level, it might be not friendly for GC. And it also brings additional copy while reading. But i have not thought of a better option now. :(

Reply:
I think we discussed it offline and agreed to some non-optimized approach.

With current `BufferBuilder` we can't avoid extra copying without resorting to accessing its underlying `memorySegment`.

As for GC, I don't know what's better here:
- allocate for each `ChannelStateByteBuffer.writeBytes`: won't escape method => no GC, but allocation cost
- allocate for each `ChannelStateStreamReader.readInto` (as it is now): short-lived - likely low GC pressure, but some allocation cost
- reuse between  `ChannelStateStreamReader.readInto` (what you proposed): memory overhead (we don't know when to clear); concurrency overhead / GC pressure with `thread-local`s (other issues?)

All in all, I think it's not a performance-critical part so we can optimize it later if needed.

Reply:
Yes, I remembered this discussion before and I agree this extra copy can not be avoided based on current codes. My previous assumption was that it would reuse the same bytes always like we did in `SpillingAdaptiveSpanningRecordDeserializer#SpanningWrapper#buffer`, to reduce GC pressure for many short-live objects.

If it is not easy to clear it, I am also fine with current way since it is not in critical path, only for recovery process. Then we can optimize it if necessary future.